### PR TITLE
[D6-C] Data Quality Warnings & KPI Anomaly Detection

### LINKED ISSUE #115: [D6] Data Quality Warnings & Anomaly Detection
# Data Quality Warnings & Anomaly Detection

Closes items 3, 6 from #111.

## Context
- **Problem**: Dashboards show stale data with no indication when ETL fails — users make bad decisions on incomplete numbers. Separately, unusual KPI values (e.g., a 40% drop in ticket medio on a Tuesday) go unnoticed until manual end-of-week review. Both are pure SQL+math problems, no LLM calls needed.
- **Worktree**: `dashboard-data-health`
- **Scope**: Two independent but thematically related features:
  1. **Data freshness banner** — warns when underlying data is stale (ETL >36h old)
  2. **KPI anomaly detection** — flags unusual KPI values with z-score analysis
- **Out of scope**: Per-widget staleness icons (future enhancement), LLM-generated anomaly explanations, proactive alerts/emails.
- **Constraints**: Non-blocking to dashboard load (both features make secondary requests). Backwards compatible (no schema changes to existing dashboard specs, new fields are optional). Read-only SQL only. Spanish UI strings.
- **Repo touchpoints**:
  - `dashboard/app/api/data-health/route.ts` (new)
  - `dashboard/app/api/anomaly-check/route.ts` (new)
  - `dashboard/components/DataFreshnessBanner.tsx` (new)
  - `dashboard/components/widgets/KpiRow.tsx` (modified — anomaly badge)
  - `dashboard/app/dashboard/[id]/page.tsx` (modified — banner integration)
  - `dashboard/lib/schema.ts` (modified — optional `anomaly_sql` on KpiItem)
  - `dashboard/components/DashboardRenderer.tsx` (modified — anomaly data fetching)
  - `dashboard/app/__tests__/` (new tests)
- **Definition of done**: Both features work end-to-end; all existing tests pass; new unit tests cover all new code; dashboard view shows freshness banner when data is stale; KPI cards show anomaly badges when `anomaly_sql` is provided and z-score exceeds threshold.
- **How is it going to be tested**: Unit tests for API routes (mocked DB), component tests for banner and anomaly badge, integration test for dashboard view with stale data scenario.

### Part A: Data Freshness Banner

- [ ] 1) Create `GET /api/data-health` API route (owner: agent)
  - **Change**: New API route that queries the `etl_watermarks` table to get the last sync timestamp per table. Returns `{ tables: { name: string, lastSync: string, isStale: boolean }[], overallStale: boolean, stalestTable: { name: string, lastSync: string } | null }`. A table is stale when `last_sync_at` is more than 36 hours ago. The query is: `SELECT table_name, last_sync_at, status FROM etl_watermarks ORDER BY last_sync_at ASC`. Note: the agent should verify the actual table and column names at implementation time — the ETL schema is defined in `etl/schema/init.sql` and the table is `etl_watermarks` with columns `table_name TEXT`, `last_sync_at TIMESTAMPTZ`, `rows_synced INTEGER`, `status TEXT`, `error_msg TEXT`.
  - **Files**: `dashboard/app/api/data-health/route.ts` (new), `dashboard/lib/db.ts` (reuse existing `query()`)
  - **Acceptance**: `curl http://localhost:4000/api/data-health` returns JSON with table freshness data. When `etl_watermarks` has a row with `last_sync_at` older than 36h, `isStale: true` for that entry and `overallStale: true`.
  - **Spec update**: mark done + update remaining tasks/context as needed

- [ ] 2) Create `DataFreshnessBanner` component (owner: agent)
  - **Change**: New React component shown at the top of the dashboard view when any referenced table is stale. Calls `GET /api/data-health` once on mount. When `overallStale` is true, displays an amber warning banner: "Los datos de [stalest table] se actualizaron por ultima vez el DD/MM/YYYY a las HH:MM. Pueden no reflejar las operaciones mas recientes." The banner is collapsible (chevron toggle) and dismissable per session (X button, uses `sessionStorage` with key `data-health-dismissed`). When all tables are fresh, the banner is not rendered. Shows a green "Datos actualizados" brief indicator that auto-hides after 3 seconds if data is fresh (optional nice-to-have).
  - **Files**: `dashboard/components/DataFreshnessBanner.tsx` (new)
  - **Acceptance**: Component renders warning when API returns stale data; does not render when data is fresh; dismiss persists within session; collapse toggles details.
  - **Spec update**: mark done

- [ ] 3) Integrate banner into dashboard view page (owner: agent)
  - **Change**: Import `DataFreshnessBanner` in `dashboard/app/dashboard/[id]/page.tsx` and render it above the `DashboardRenderer` component (below the top toolbar, above the dashboard content). The banner fetches independently — it does not block dashboard rendering. Also add the banner to `dashboard/app/dashboard/new/page.tsx` (the creation page shows it after the first dashboard is generated).
  - **Files**: `dashboard/app/dashboard/[id]/page.tsx`, `dashboard/app/dashboard/new/page.tsx`
  - **Acceptance**: When viewing any dashboard, the freshness banner appears at the top if data is stale. Dashboard content loads independently (banner does not block widgets).
  - **Spec update**: mark done

- [ ] 4) Unit tests for data freshness feature (owner: agent)
  - **Change**: Tests for the `/api/data-health` route (mock `query()` from `@/lib/db` to return fresh/stale watermark rows; verify response shape). Component tests for `DataFreshnessBanner` (mock fetch, verify banner renders/hides, verify dismiss behavior via sessionStorage mock).
  - **Files**: `dashboard/app/__tests__/data-health-api.test.ts` (new), `dashboard/app/__tests__/data-freshness-banner.test.tsx` (new)
  - **Acceptance**: All new tests pass with `npm test`. Coverage includes: fresh data (no banner), stale data (banner shown), dismiss (banner hidden), API error (graceful fallback — no banner, no crash).
  - **Spec update**: mark done

### Part B: KPI Anomaly Detection

- [ ] 5) Add `anomaly_sql` optional field to KpiItem schema (owner: agent)
  - **Change**: Add `anomaly_sql?: string` to the `KpiItemSchema` in `dashboard/lib/schema.ts`. This SQL, when provided, returns N rows of historical values (e.g., same KPI metric for the last 8 comparable periods). The frontend uses these to compute a z-score client-side. This approach avoids parsing arbitrary SQL server-side — the LLM generates the historical query at dashboard creation time (same pattern as `trend_sql`). The field is optional for full backwards compatibility.
  - **Files**: `dashboard/lib/schema.ts`
  - **Acceptance**: Existing dashboard specs without `anomaly_sql` still validate. New specs with `anomaly_sql` set also validate. `anomaly_sql: ""` is rejected (min length 1).
  - **Spec update**: mark done

- [ ] 6) Create `POST /api/anomaly-check` API route (owner: agent)
  - **Change**: New API route that accepts `{ sql: string }` where `sql` is the `anomaly_sql` query. It executes the SQL (which returns N rows of historical values as single-column numeric results), computes the z-score of the first value (current period) against the remaining values (historical), and returns `{ isAnomaly: boolean, currentValue: number, mean: number, stddev: number, zScore: number, direction: "high" | "low" | "normal", explanation: string }`. Anomaly threshold: `|z-score| > 2.0`. The `explanation` is a template string in Spanish: e.g., "El valor actual (18.40) esta un 32% por debajo de la media de los ultimos 7 periodos (27.10)." If fewer than 4 historical values, return `{ isAnomaly: false }` (insufficient data). Alternative: the z-score computation can be done entirely client-side in the KpiRow component without a dedicated API route — the agent should evaluate which approach is simpler and choose accordingly. If client-side, skip this task and do the computation in task 7.
  - **Files**: `dashboard/app/api/anomaly-check/route.ts` (new) OR inline in `dashboard/components/widgets/KpiRow.tsx`
  - **Acceptance**: Given historical values `[100, 98, 102, 99, 101, 50]` where 50 is the current period: returns `isAnomaly: true, direction: "low"`. Given `[100, 98, 102, 99, 101, 100]`: returns `isAnomaly: false`.
  - **Spec update**: mark done + note which approach was chosen (API route vs client-side)

- [ ] 7) Add anomaly badge to KpiRow component (owner: agent)
  - **Change**: For each KPI item that has `anomaly_sql` set, fetch the anomaly data (either via `/api/anomaly-check` or by fetching the `anomaly_sql` directly via `/api/query` and computing client-side). The fetch happens AFTER the main KPI data loads (non-blocking, secondary request — same pattern as `trendData` in `DashboardRenderer.tsx`). When `isAnomaly: true`, display a small amber (direction=low) or red (direction=high, for metrics where high is bad) badge next to the KPI value showing "Valor inusual" with a tooltip (title attribute) containing the explanation. Badge styling: `bg-amber-100 text-amber-700 dark:bg-amber-900/30 dark:text-amber-400` (similar to the existing `TrendBadge` pattern in `KpiRow.tsx`). When anomaly data is loading, show nothing extra (no skeleton for anomaly — it's supplementary). When anomaly check fails, show nothing (graceful degradation).
  - **Files**: `dashboard/components/widgets/KpiRow.tsx`, `dashboard/components/DashboardRenderer.tsx` (add anomaly fetching alongside trend fetching in `fetchAll` and `retryWidget`)
  - **Acceptance**: KPI card with `anomaly_sql` shows "Valor inusual" badge when z-score > 2. KPI card without `anomaly_sql` renders exactly as before (no visual change). Anomaly fetch failure does not affect KPI rendering.
  - **Spec update**: mark done

- [ ] 8) Update LLM prompt to generate `anomaly_sql` (owner: agent)
  - **Change**: Update the dashboard generation system prompt in `dashboard/lib/knowledge.ts` (or wherever the LLM prompt is constructed) to instruct the LLM to optionally include `anomaly_sql` on KPI items. The instruction should say: "For each KPI item, optionally add an `anomaly_sql` field containing a SQL query that returns the same metric for the last 8 comparable periods (e.g., same weekday for daily metrics, same week for weekly). The first row should be the current period value, followed by historical values in descending chronological order. Only add `anomaly_sql` for metrics where anomaly detection adds value (sales, ticket medio, margin) — skip for static counts or configuration values." Also update the spec format documentation in the prompt to show the new field.
  - **Files**: `dashboard/lib/knowledge.ts` or `dashboard/lib/prompts.ts` (wherever the system prompt lives)
  - **Acceptance**: New dashboards generated by the LLM include `anomaly_sql` on appropriate KPI items. The SQL is valid PostgreSQL against `ps_*` tables.
  - **Spec update**: mark done

- [ ] 9) Unit tests for anomaly detection feature (owner: agent)
  - **Change**: Tests for z-score computation (pure function, edge cases: insufficient data, zero stddev, normal values, anomalous values). Component tests for KpiRow with anomaly badge (mock fetch, verify badge appears/hides). If API route approach: tests for `/api/anomaly-check` route.
  - **Files**: `dashboard/app/__tests__/anomaly-check.test.ts` (new), `dashboard/app/__tests__/kpi-anomaly.test.tsx` (new)
  - **Acceptance**: All new tests pass. Z-score edge cases covered (empty array, single value, all identical values, clear anomaly, borderline values).
  - **Spec update**: mark done

### Part C: Finalization

- [ ] 10) Run all checks and fix issues (owner: agent)
  - **Change**: Run all tests, linting, type-checking; fix any failures.
  - **Files**: any files with issues
  - **Acceptance**: `cd dashboard && npm run lint && npx tsc --noEmit && npm test` all pass. No regressions in existing tests.
  - **Spec update**: mark done

- [ ] 11) Review cycle (owner: agent)
  - **Change**: Request Copilot review, address all feedback, re-request until no new feedback.
  - **How**: `gh pr create` then request Copilot review via REST API: `gh api repos/{owner}/{repo}/pulls/{PR#}/requested_reviewers --method POST -f 'reviewers[]=copilot-pull-request-reviewer[bot]'`. Poll for review: `gh api repos/{owner}/{repo}/pulls/{PR#}/reviews --jq '[.[] | {state, user: .user.login, body}]'`. Address all comments with inline replies. Re-request after each round.
  - **Acceptance**: Copilot review shows no unresolved comments.
  - **Spec update**: mark done

- [ ] 12) Create commit (owner: agent)
  - **Change**: Stage all changes and create a descriptive commit.
  - **Files**: none (git operation)
  - **Acceptance**: `git status` shows clean working tree; `git log -1` shows the new commit.
  - **Spec update**: mark done

### ETL Watermarks Table Schema (from `etl/schema/init.sql`)
```sql
CREATE TABLE IF NOT EXISTS etl_watermarks (
    table_name   TEXT        PRIMARY KEY,
    last_sync_at TIMESTAMPTZ NOT NULL,
    rows_synced  INTEGER,
    status       TEXT        DEFAULT 'ok',
    error_msg    TEXT
);
```
The `set_watermark()` function in `etl/db/postgres.py` also writes an `updated_at` column (added via upsert ON CONFLICT). The agent should verify the full column set at implementation time.

### Existing Patterns to Follow
- **Trend data fetching**: `DashboardRenderer.tsx` already fetches `trendData` alongside main KPI data in `fetchAll()` and `retryWidget()`. The anomaly data should follow the exact same pattern — fetch in parallel, store in `WidgetState`, pass to `KpiRow`.
- **TrendBadge component**: `KpiRow.tsx` has a `TrendBadge` component with the exact styling pattern to reuse for the anomaly badge.
- **API route pattern**: `dashboard/app/api/query/route.ts` shows the standard error handling, SQL validation, and response format.
- **Schema extension**: `KpiItemSchema` already has optional `trend_sql` — adding `anomaly_sql` follows the same pattern.
- **Read-only SQL**: All SQL must pass `validateReadOnly()` from `dashboard/lib/db.ts`.

### Key Design Decisions
- **`anomaly_sql` over server-side SQL parsing**: Instead of having the API parse arbitrary KPI SQL to generate historical queries, the LLM generates the `anomaly_sql` at dashboard creation time (same as `trend_sql`). This is simpler, more reliable, and gives the LLM full control over the comparison logic.
- **Z-score threshold of 2.0**: Standard statistical threshold (~95% confidence). Can be made configurable later.
- **36-hour staleness threshold**: ETL runs nightly at 02:00. Data older than 36h means at least one ETL run has been missed.
- **Session-scoped dismiss**: Banner dismiss resets on new browser session. Not persisted to DB (no user auth system).

### PR DESCRIPTION
**Pull request by alvarolobato**
## Summary

- **Data freshness banner**: `GET /api/data-health` queries `etl_watermarks` for staleness (36h threshold). `DataFreshnessBanner` component shows an amber warning with collapse/dismiss (session-scoped via `sessionStorage`). Integrated into the dashboard view and new dashboard pages. Gracefully degrades when the table is missing.
- **KPI anomaly detection**: `POST /api/anomaly-check` computes z-score from historical SQL values (threshold 2.0). `anomaly_sql` optional field added to `KpiItemSchema`. `KpiRow` computes z-score client-side and shows an amber `AnomalyBadge` ("Valor inusual") with a Spanish explanation tooltip. `DashboardRenderer` fetches `anomaly_sql` data alongside `trend_sql` in `fetchAll` and `retryWidget`.
- **LLM prompt updated**: instructs the LLM to optionally add `anomaly_sql` on appropriate KPI items (sales totals, ticket medio, margin).
- **37 new tests** across 4 test files; all 371 tests pass.

## Test plan

- [x] `cd dashboard && npm test` — all 371 tests pass
- [x] `DataFreshnessBanner` renders warning when API returns stale data; hidden when fresh; dismiss persists in session; collapse toggles detail list
- [x] `GET /api/data-health` returns empty gracefully when `etl_watermarks` is missing (42P01), on connection error, or on unexpected error
- [x] `POST /api/anomaly-check` returns `isAnomaly: true` for [50, 100, 98, 102, 99, 101] and `false` for [100, 98, 102, 99, 101, 100]
- [x] `KpiRow` shows anomaly badge only when z-score > 2.0 and `anomaly_sql` is set; no badge for normal values or missing data
- [x] Lint and type-check pass (pre-existing errors in `ErrorDisplay.test.tsx` and `ChatSidebar.test.tsx` unchanged)

Closes #115

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: dashboard/app/api/anomaly-check/route.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    const values: number[] = [];
+    for (const row of result.rows) {
+      const raw = row[0];
+      if (raw !== null && raw !== undefined) {
+        const num = Number(raw);
+        if (!isNaN(num)) {
+          values.push(num);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (alvarolobato) [reply dashboard/app/api/anomaly-check/route.ts]: Fixed in 8cbb898. The POST handler in `anomaly-check/route.ts` now validates row 0 separately — returns `{ isAnomaly: false }` if null/non-numeric. Historical rows are then filtered for nulls. The `computeAnomaly` pure function receives a correctly aligned `[currentNum, ...historical]` array.

File: dashboard/components/widgets/KpiRow.tsx [hunk 2]
```diff
@@ -19,6 +19,74 @@ interface KpiRowProps {
...
+    zScore > ANOMALY_Z_THRESHOLD ? "high" : "low";
+
+  const pctChange =
+    mean !== 0 ? ((currentValue - mean) / Math.abs(mean)) * 100 : 0;
+  const absPct = Math.abs(pctChange).toFixed(0);
+  const dirText = direction === "high" ? "por encima" : "por debajo";
+  const fmt = (v: number) =>
+    v.toLocaleString("es-ES", { maximumFractionDigits: 2 });
+
+  const explanation = `El valor actual (${fmt(currentValue)}) está un ${absPct}% ${dirText} de la media de los últimos ${n} periodos (${fmt(mean)}).`;
```
[the remark this answers is not in the document]
Comment (alvarolobato) [reply dashboard/components/widgets/KpiRow.tsx]: Fixed in 73126fd. All occurrences of "periodos" in KpiRow.tsx replaced with "períodos".

File: dashboard/app/api/anomaly-check/route.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+      : "normal";
+
+  const pctChange =
+    mean !== 0 ? ((currentValue - mean) / Math.abs(mean)) * 100 : 0;
+  const absPct = Math.abs(pctChange).toFixed(0);
+  const dirText = direction === "high" ? "por encima" : "por debajo";
+
+  const explanation =
+    direction !== "normal"
+      ? `El valor actual (${formatNum(currentValue)}) está un ${absPct}% ${dirText} de la media de los últimos ${n} períodos (${formatNum(mean)}).`
```
[the remark this answers is not in the document]
Comment (alvarolobato) [reply dashboard/app/api/anomaly-check/route.ts]: Fixed in 6419001. Same fix applied in the API route: when mean !== 0, show percentage; when mean === 0, show absolute difference with "diferencia absoluta".
